ion-hash/tests/ion_hash_tests.rs

```diff
@@ -1,71 +1,96 @@
+use digest::{consts::U256, generic_array::GenericArray, Digest, Output};
+use ion_hash::{self, IonHasher};
 use ion_rs::result::IonResult;
 use ion_rs::value::loader::{loader, Loader};
 use ion_rs::value::*;
-use std::{
-    fs::read,
-    sync::{Arc, Mutex},
-};
+use std::fs::read;
 
 #[derive(Default, Clone)]
 struct TestDigest {
-    inner: Arc<Mutex<TestDigestInner>>,
+    updates: GenericArray<u8, U256>,
+    position: usize,
 }
 
-#[derive(Default)]
-struct TestDigestInner {
-    updates: Vec<u8>,
-    digest: Vec<u8>,
-}
+impl Digest for TestDigest {
+    // Pick a number bigger than the biggest test case digest expectation.
+    type OutputSize = U256;
 
-impl TestDigest {
-    fn inspect(&self) -> (Vec<u8>, Vec<u8>) {
-        let inner = self.inner.lock().unwrap();
-        (inner.updates.clone(), inner.digest.clone())
+    fn new() -> Self {
+        Self {
+            updates: GenericArray::default(),
+            position: 0,
+        }
     }
-}
 
-impl IonHashDigest for TestDigest {
     fn update(&mut self, bytes: impl AsRef<[u8]>) {
-        let mut inner = self.inner.lock().unwrap();
         for b in bytes.as_ref() {
-            inner.updates.push(*b);
-            inner.digest.push(*b);
+            self.updates[self.position] = *b;
+            self.position += 1;
         }
     }
 
-    fn finalize(self) -> Vec<u8> {
-        let inner = self.inner.lock().unwrap();
-        inner.digest.clone()
+    fn chain(self, _data: impl AsRef<[u8]>) -> Self
+    where
+        Self: Sized,
+    {
+        todo!()
+    }
+
+    fn finalize(self) -> Output<Self> {
+        self.updates
+    }
+
+    fn finalize_reset(&mut self) -> Output<Self> {
+        let output = self.updates;
+        self.updates = GenericArray::default();
+        self.position = 0;
+        output
+    }
+
+    fn reset(&mut self) {
+        self.updates = GenericArray::default();
+        self.position = 0;
+    }
+
+    fn output_size() -> usize {
+        256
+    }
+
+    fn digest(data: &[u8]) -> Output<Self> {
+        let mut myself = Self::new();
+        myself.update(data);
+        myself.finalize()
     }
 }
 
 #[test]
-fn the_first_test() -> IonResult<()> {
-    let example = br#"
-{ ion:null,           expect:{ identity:(update::(0x0b) update::(0x0f) update::(0x0e) digest::(0x0b 0x0f 0x0e)),
-                                    md5:(digest::(0x0f 0x50 0xc5 0xe5 0xe8 0x77 0xb4 0x45 0x1a 0xa9 0xfe 0x77 0xc3 0x76 0xcd 0xe4)) } }"#;
-    let elems = loader().load_all(example)?;
+fn ion_hash_tests() -> IonResult<()> {
+    test_file("tests/ion_hash_tests.ion")
+}
+
+fn test_file(file_name: &str) -> IonResult<()> {
+    let data = read(file_name)?;
+    let elems = loader().load_all(&data)?;
     test_all(elems)
 }
 
 fn test_all<E: Element>(elems: Vec<E>) -> IonResult<()> {
-    for case in elems {
+    for (i, case) in elems.iter().enumerate() {
         let case = case.as_struct().expect("test cases are structs");
```

Comment:
Lack of table driven testing for Rust's testing framework makes me a sad :panda_face:.

Reply:
yup!